joern-cli/frontends/rubysrc2cpg/src/main/scala/io/joern/rubysrc2cpg/RubySrc2Cpg.scala

```diff
@@ -1,26 +1,58 @@
 package io.joern.rubysrc2cpg
 
-import io.joern.rubysrc2cpg.passes.{AstCreationPass, ConfigPass}
+import io.joern.rubysrc2cpg.passes.{AstCreationPass, AstPackagePass, PackageResolverPass, ConfigPass}
+import io.joern.rubysrc2cpg.utils.PackageTable
 import io.joern.x2cpg.X2Cpg.withNewEmptyCpg
 import io.joern.x2cpg.X2CpgFrontend
+import io.joern.x2cpg.datastructures.Global
 import io.joern.x2cpg.passes.frontend.{MetaDataPass, TypeNodePass}
 import io.shiftleft.codepropertygraph.Cpg
 import io.shiftleft.codepropertygraph.generated.Languages
 import org.slf4j.LoggerFactory
 
-import scala.util.Try
+import java.nio.file.{Files, Paths}
+import scala.sys.process._
+import scala.util.{Failure, Success, Try}
 
 class RubySrc2Cpg extends X2CpgFrontend[Config] {
 
   private val logger = LoggerFactory.getLogger(this.getClass)
 
+  val global = new Global()
+
   override def createCpg(config: Config): Try[Cpg] = {
     withNewEmptyCpg(config.outputPath, config: Config) { (cpg, config) =>
+      val packageTableInfo = new PackageTable()
       new MetaDataPass(cpg, Languages.RUBYSRC, config.inputPath).createAndApply()
       new ConfigPass(cpg, config.inputPath).createAndApply()
-      val astCreationPass = new AstCreationPass(config.inputPath, cpg)
+      val astCreationPass = new AstCreationPass(config.inputPath, cpg, global, packageTableInfo)
       astCreationPass.createAndApply()
+      downloadDependency(config.inputPath)
+      new AstPackagePass(cpg, getGemEnv(), global, packageTableInfo, config.inputPath).createAndApply()
+      new PackageResolverPass(cpg, packageTableInfo).createAndApply()
       new TypeNodePass(astCreationPass.allUsedTypes(), cpg).createAndApply()
     }
   }
+
+  private def getGemEnv(): List[String] = {
+    // fetch gem path where dependency file present
+    Try("gem env gempath".!!) match {
+      case Success(gemPath) =>
+        gemPath.split(":").toList
+      case Failure(exception) =>
+        logger.error(s"Error While fetching gem path: ${exception.getMessage}")
+        List.empty
+    }
+  }
+
+  private def downloadDependency(inputPath: String): Unit = {
```

Comment:
This worries me: I don't think we should be messing with the user's environment in any capacity when scanning. At best, I'd put this behind a feature flag. Ideally, we would live without it.

Reply:
If I understand things through, we're downloading and scanning the dependencies together with the project in order to resolve full-names, right? We have a similar challenge in JS with imports. In Ruby it's arguably worse since `require`/`include` is but a "statement" (in JS we'd have some variable assigned to its result or some aliasing).
cc: @DavidBakerEffendi, @max-leuthaeuser  

In this approach, couldn't it happen we also need to download and scan the original dependencies' dependencies and so on until closure?

I'd frame this approach as an incremental improvement available through some feature flag, and assume we haven't scanned the dependencies by default.

Reply:
+1 to making it disabled by default, we might introduce a CWE through Joern otherwise. And yes, there would also result in some kind of recursive download situation. Ideally, there is an API we can use to gather all the dependencies and their transitive dependencies to do 1 download and force Joern into a whole-program analysis.

But beware, whole-program analysis is slow and resource intensive.

Reply:
@ankit-privado As a tangent, I'm working on an API server to fetch these infos by wrapping the joern frontends. There might some opportunities to collaborate since your optional pass could use this service that runs locally or remotely or use a cli.

Reply:
Done @DavidBakerEffendi @xavierpinho, added flag to download the dependency, by default it will not download anything.